howto_io.md

```diff
@@ -0,0 +1,268 @@
+actual sound data:
+
+```
+chunk_id         "RIFF"
+chunk_size       4-byte unsigned little-endian integer
+format           "WAVE"
+fmt_id           "fmt "
+fmt_size         4-byte unsigned little-endian integer
+audio_fmt        2-byte unsigned little-endian integer
+num_channels     2-byte unsigned little-endian integer
+sample_rate      4-byte unsigned little-endian integer
+byte_rate        4-byte unsigned little-endian integer
+block_align      2-byte unsigned little-endian integer
+bits_per_sample  2-byte unsigned little-endian integer
+data_id          "data"
+data_size        4-byte unsigned little-endian integer
+```
+The `.wav` file header as a NumPy structured dtype:
+```
+wav_header_dtype = np.dtype([
+    ("chunk_id", (bytes, 4)), # flexible-sized scalar type, item size 4
+    ("chunk_size", "<u4"),    # little-endian unsigned 32-bit integer
+    ("format", "S4"),         # 4-byte string
+    ("fmt_id", "S4"),
+    ("fmt_size", "<u4"),
+    ("audio_fmt", "<u2"),     #
+    ("num_channels", "<u2"),  # .. more of the same ...
+    ("sample_rate", "<u4"),   #
+    ("byte_rate", "<u4"),
+    ("block_align", "<u2"),
+    ("bits_per_sample", "<u2"),
+    ("data_id", "S4"),
+    ("data_size", "u4"),
+    #
+    # the sound data itself cannot be represented here:
+    # it does not have a fixed size
+   ])
+
+data = np.load(f,dtype=wave_header_dtype)
+```
+Credit: [Pauli Virtanen](http://scipy-lectures.org/advanced/advanced_numpy/index.html)
+
+<a name="large_arrays">
+
+## Write or read large arrays
+
+Arrays too large to fit in memory can be treated like ordinary in-memory arrays using
+[numpy.mmap](https://numpy.org/doc/stable/reference/generated/numpy.memmap.html).
+```
+import numpy as np
+array = np.memmap("mydata/myarray.arr", mode="r",
+                  dtype=np.int16, shape=(1024, 1024))
+```
+The simplicity also means it's possible to have access patterns that don't
+match `memmap`'s buffering and therefore access disk repeatedly.
+[Zarr](https://zarr.readthedocs.io/en/stable/) and the similar
+[HDF5](https://portal.hdfgroup.org/display/support) formats allow you to
+manage buffering optimally for your program's access pattern.
+
+Itamar Turner-Trauring in
+[pythonspeed.com](https://pythonspeed.com/articles/mmap-vs-zarr-hdf5/) gives
+more details and describes tradeoffs among memmap, Zarr, and HDF5.
+
+* For **HDF5**, use [h5py](https://www.h5py.org/) or [PyTables](https://www.pytables.org/).
+* For **Zarr**, see [here](https://zarr.readthedocs.io/en/stable/tutorial.html#reading-and-writing-data).
+* For **NetCDF**, use [scipy.io.netcdf_file](https://docs.scipy.org/doc/scipy/reference/generated/scipy.io.netcdf_file.html).
+
+## Write files for reading by other (non-NumPy) tools
+
+Formats for exchanging data with other tools include HDF5, Zarr, and NetCDF.
+
+## Write or read a JSON file
+
+NumPy arrays are not directly [JSON serializable](https://github.com/numpy/numpy/issues/12481).

```

Comment:
Might be helpful to mention "since the JSON format only defines ints and floats, while NumPy has a rich set of dtypes"

Reply:
NumPy converts simpler types in other contexts, so the reader who sees this explanation may be skeptical. It's got to be more than this.  We do link to the issue, and having read it it just looks like a nontrivial problem NumPy hasn't chosen to solve.